### DISCUSSION TITLE
Possible to use NamedTemporaryFile instead of SpooledTemporaryFile?

### DISCUSSION THREAD

**Discussion by makupi** [fastapi/fastapi#9102] [Questions]
### First check

* [ x ] I added a very descriptive title to this issue.
* [ x ] I used the GitHub search to find a similar issue and didn't find it.
* [ x ] I searched the FastAPI documentation, with the integrated search.
* [ x ] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [ x ] I already read and followed all the tutorial in the docs and didn't find an answer.
* [ x ] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [ x ] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [ x ] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [ x ] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

```Python
from fastapi import FastAPI, UploadFile, File

app = FastAPI()


@app.post("/upload")
def upload_file(file: UploadFile = File(...)):
    return {"temporaryFileName": file.file.name}
```

### Description

I am running FastAPI on a very small system, with only a little under 1GB left for storage. Due to this limitation I'm running into an Issue where when uploading large files (in this case a Firmware Update) the temporary file already takes up half of the empty storage and I'm therefore unable to store a copy that I can actually work with. Therefore it would make more sense to have a TemporaryFile that I can work with directly.     

Is there any way to use a NamedTemporaryFile (as I understand these can be accessed by other processes as long as they exist?) instead of a SpooledTemporaryFile? Or would I have to stream the received data and store it in some file myself?

### Environment

* OS: [ Linux armhf ]:
* FastAPI Version [ 0.60.0 ]:

* Python version: 3.7.8

**Comment by Arpafaucon** [comment]
I'm having a similar issue, where I need to better control where uploaded files end up.
In our case, we know that some of the routes are expected to receive massive files (~10Gb); a plain `UploadFile` will create a `SpooledTemporaryFile` in `/tmp` (which is a `tmpfs` on our server), and overflow the RAM.

In my case, being able to set the `dir` argument of `SpooledTemporaryFile` would solve the issue. However this parameter is not accessible at fastapi level

@makupi I am not very confident FastAPI can do something, this seems like a `starlette` issue.
Some issues worth reading, you may find something that fits your need: 
- https://github.com/encode/starlette/issues/697 
- https://github.com/tiangolo/fastapi/issues/390#issuecomment-511766926
- https://github.com/encode/starlette/issues/388

**Reply by CarliJoy** [reply]
~~@makupi~~ @Arpafaucon  (sorry for the wrong ping) you can change the default location of tempory by overwriting `tempfile.tempdir = "/home/myapp"`.

This way you won't run out of memory.

You still would have at least 20Gb on this drive as there is no way to access the file object of a SpooledTemporaryFile in Linux, so you will have to write a copy yet again on the drive.

**Comment by JuantAldea** [comment]
I'm facing the same issue. My /tmp folder a temfs and I'm not so invested on whole re-pointing tempfile.tempdir folder for the whole application.

Moreover,  as @CarliJoy pointed out, there's a moment where both, the tempfile and the final-upload coexist on the device, despite having TMP_DIR and the final file location stored the same filesystem.
